util/extsort/disk_sorter.go

```diff
@@ -131,29 +891,330 @@ func (d *DiskSorter) Sort(ctx context.Context) error {
+		})
+		for i := minIntervalIndex; i < maxIntervalIndex; i++ {
+			if intervals[i].depth >= d.opts.CompactionThreshold {
+				files = append(files, file)
+				break
+			}
+		}
+	}
+	d.opts.Logger.Info(
+		"max overlap depth exceeds the compaction threshold, pick files to compact",
+		zap.Int("maxDepth", maxDepth),
+		zap.Int("threshold", d.opts.CompactionThreshold),
+		zap.Int("fileCount", len(files)),
+	)
+	return files
+}
+
+func splitCompactionFiles(files []*fileMetadata, maxCompactionDepth int) [][]*fileMetadata {
+	// Split files into non-overlapping groups.
+	slices.SortFunc(files, func(a, b *fileMetadata) bool {
+		return bytes.Compare(a.startKey, b.startKey) < 0
+	})
+	var groups [][]*fileMetadata
+	curGroup := []*fileMetadata{files[0]}
+	maxEndKey := files[0].endKey
+	for _, file := range files[1:] {
+		if bytes.Compare(file.startKey, maxEndKey) > 0 {
+			groups = append(groups, curGroup)
+			curGroup = []*fileMetadata{file}
+		} else {
+			curGroup = append(curGroup, file)
+		}
+		if bytes.Compare(file.endKey, maxEndKey) > 0 {
+			maxEndKey = file.endKey
+		}
+	}
+	if len(curGroup) > 0 {
+		groups = append(groups, curGroup)
+	}
+
+	var finalGroups [][]*fileMetadata
+	// Compact each group of files.
+	for _, group := range groups {
+		numSubGroups := (len(group)-1)/maxCompactionDepth + 1
+		subGroupSize := (len(group)-1)/numSubGroups + 1
+		for i := 0; i < len(group); i += subGroupSize {
+			j := i + subGroupSize
+			if j > len(group) {
+				j = len(group)
+			}
+			finalGroups = append(finalGroups, group[i:j])
+		}
+	}
+	return finalGroups
+}
+
+type compaction struct {
+	startKey     []byte
+	endKey       []byte
+	orderedFiles []*fileMetadata // ordered by start key
+}
+
+func buildCompactions(files []*fileMetadata, maxCompactionSize int) []*compaction {
+	var startKey, endKey []byte
+	buckets := make([]kvStatsBucket, 0, len(files))
+	for _, file := range files {
+		buckets = append(buckets, file.kvStats.Histogram...)
+		if startKey == nil || bytes.Compare(file.startKey, startKey) < 0 {
+			startKey = file.startKey
+		}
+		if endKey == nil || bytes.Compare(file.endKey, endKey) > 0 {
+			endKey = file.endKey
+		}
+	}
+	slices.SortFunc(buckets, func(a, b kvStatsBucket) bool {
+		return bytes.Compare(a.UpperBound, b.UpperBound) < 0
+	})
+
+	var (
+		kvSize      int
+		compactions []*compaction
+	)
+	for _, bucket := range buckets {
+		kvSize += bucket.Size
+		if kvSize >= maxCompactionSize {
+			compactions = append(compactions, &compaction{
+				startKey: startKey,
+				endKey:   bucket.UpperBound,
+			})
+			startKey = bucket.UpperBound
+			kvSize = 0
+		}
+	}
+	if bytes.Compare(startKey, endKey) < 0 {
+		compactions = append(compactions, &compaction{
+			startKey: startKey,
+			endKey:   endKey,
+		})
+	}
+
+	for _, c := range compactions {
+		for _, file := range files {
+			if !(bytes.Compare(file.endKey, c.startKey) <= 0 ||
+				bytes.Compare(file.startKey, c.endKey) >= 0) {
+				c.orderedFiles = append(c.orderedFiles, file)
```

Comment:
IMO it's easier to understand to rename `orderedFiles` to overlapping files and change if condition to positively checking overlap. "ordered by start key" can only be mentioned in comments

Reply:
Renamed in https://github.com/pingcap/tidb/pull/44185/commits/7240482fd168f8fceca7e8c58f54f563535800c9.

But for the if condition, I think the negative version is more easy to understand.